File: LeetCode/java/267_Palindrome_Permutation_II.java
import java.util.*;

class Solution {
    Map<Character, Integer> map = new HashMap<>();
    List<Character> chars = new ArrayList<>();
    List<Character> list = new ArrayList<>();
    boolean[] visit;
    Set<String> st = new HashSet<>();

    public List<String> generatePalindromes(String s) {
        List<String> answer = new ArrayList<>();

        for (int i = 0; i < s.length(); i++) {
            map.put(s.charAt(i), map.getOrDefault(s.charAt(i), 0) + 1);
        }

        if (!validateMap()) {
            return answer;
        }

        char oddCntChar = '*';
        for (char key : map.keySet()) {
            if (map.get(key) % 2 != 0) {
                oddCntChar = key;
            }

            for (int i = 0; i < map.get(key) / 2; i++) {
                chars.add(key);
            }
        }

        visit = new boolean[chars.size()];

        dfs(0, s, oddCntChar);

        for (String ss : st) {
            answer.add(ss);
        }

        return answer;
    }

    private void dfs(int cnt, String s, char oddChar) {
        if (cnt == chars.size()) {
            StringBuilder temp = new StringBuilder();
            StringBuilder result = new StringBuilder();

            for (int i = 0; i < list.size(); i++) {
                temp.append(list.get(i));
            }

            StringBuilder rev = new StringBuilder(temp.toString());
            rev.reverse();

            if (oddChar == '*') {
                result.append(temp.toString());
                result.append(rev.toString());
            } else {
                result.append(temp.toString());
                result.append(oddChar);
                result.append(rev.toString());
            }

            st.add(result.toString());

            return;
        }

        for (int i = 0; i < chars.size(); i++) {
            if (visit[i] == true) {
                continue;
            }

            visit[i] = true;
            list.add(chars.get(i));
            dfs(cnt + 1, s, oddChar);
            list.remove(list.size() - 1);
            visit[i] = false;
        }
    }

    private boolean validateMap() {
        boolean flag = false;

        for (char key : map.keySet()) {
            if (map.get(key) % 2 != 0) {
                if (!flag) {
                    flag = true;
                } else {
                    return false;
                }
            }
        }

        return true;
    }
}

File: LeetCode/cpp/267_Palindrome_Permutation_II.cpp
class Solution {
public:
	unordered_map<char, int> um;
	vector<char> chars, v;
	vector<bool> visit;
	set<string> st;

	bool validateMap() {
		bool flag = false;

		for (pair<char, int> p : um) {
			if (p.second % 2 != 0) {
				if (!flag) {
					flag = true;
				}
				else {
					return false;
				}
			}
		}

		return true;
	}

	void dfs(int cnt, string s, char oddChar) {
		if (cnt == chars.size()) {
			string temp = "", rev = "", result = "";

			for (int i = 0; i < v.size(); i++) {
				temp += v[i];
			}

			rev = temp;
			reverse(rev.begin(), rev.end());

			if (oddChar == '*') {
				result = temp + rev;
			}
			else {
				result = temp + oddChar + rev;
			}

			st.insert(result);

			return;
		}

		for (int i = 0; i < chars.size(); i++) {
			if (visit[i] == true) {
				continue;
			}

			visit[i] = true;
			v.push_back(chars[i]);
			dfs(cnt + 1, s, oddChar);
			v.pop_back();
			visit[i] = false;
		}
	}

	vector<string> generatePalindromes(string s) {
		vector<string> answer;

		for (int i = 0; i < s.length(); i++) {
			um[s[i]]++;
		}

		if (!validateMap()) {
			return answer;
		}

		char oddCntChar = '*';
		for (pair<char, int> p : um) {
			if (p.second % 2 != 0) {
				oddCntChar = p.first;
			}

			for (int i = 0; i < p.second / 2; i++) {
				chars.push_back(p.first);
			}
		}

		visit.assign(chars.size(), false);

		dfs(0, s, oddCntChar);

		for (string ss : st) {
			answer.push_back(ss);
		}

		return answer;
	}
};